Products/GenericSetup/tool.py

```diff
@@ -1166,7 +1178,16 @@ def _runImportStepsFromContext(self,
 
         detect_steps = steps is None
 
-        for profile_id in chain:
+        last_index = len(chain) - 1
+        for index, profile_id in enumerate(chain):
+            if not always_apply_profiles and index != last_index:
+                # Check if dependency profile was already applied.
+                if self.getLastVersionForProfile(profile_id) != 'unknown':
+                    # Profile was already applied.
+                    # Maybe apply upgrade steps, if any, otherwise continue.
+                    if upgrade_dependencies:
+                        self.upgradeProfile(profile_id)
+                    continue
             context = self._getImportContext(profile_id, purge_old, archive)
             self.applyContext(context)
```

Comment:
@Themanwithoutaplan Regarding the last comment, I think you are just missing that these two lines are still inside the loop. The should make the logic clearer.